mne/time_frequency/tests/test_spectrum.py

```diff
@@ -0,0 +1,188 @@
+from functools import partial
+
+import numpy as np
+import pytest
+from numpy.testing import assert_array_equal
+
+from mne.time_frequency import read_spectrum
+from mne.time_frequency.multitaper import _psd_from_mt
+from mne.utils import requires_h5py, requires_pandas
+
+
+def test_spectrum_errors(raw):
+    """Test for expected errors in the .compute_psd() method."""
+    with pytest.raises(ValueError, match='must not exceed ½ the sampling'):
+        raw.compute_psd(fmax=raw.info['sfreq'] * 0.51)
+    with pytest.raises(TypeError, match='unexpected keyword argument foo for'):
+        raw.compute_psd(foo=None)
+    with pytest.raises(TypeError, match='keyword arguments foo, bar for'):
+        raw.compute_psd(foo=None, bar=None)
+
+
+@pytest.mark.parametrize('method', ('welch', 'multitaper'))
+@pytest.mark.parametrize(
+    ('fmin, fmax, tmin, tmax, picks, proj, n_fft, n_overlap, n_per_seg, '
+     'average, window, bandwidth, adaptive, low_bias, normalization'),
+    [[0, np.inf, None, None, None, False, 256, 0, None,
+      'mean', 'hamming', None, False, True, 'length'],  # defaults
+     [5, 50, 1, 6, 'grad', True, 128, 8, 32,
+      'median', 'triang', 10, True, False, 'full']  # non-defaults
+     ]
+)
+def test_spectrum_params(method, fmin, fmax, tmin, tmax, picks, proj, n_fft,
+                         n_overlap, n_per_seg, average, window, bandwidth,
+                         adaptive, low_bias, normalization, raw):
+    """Test valid parameter combinations in the .compute_psd() method."""
+    kwargs = dict(method=method, fmin=fmin, fmax=fmax, tmin=tmin, tmax=tmax,
+                  picks=picks, proj=proj)
+    if method == 'welch':
+        kwargs.update(n_fft=n_fft, n_overlap=n_overlap, n_per_seg=n_per_seg,
+                      average=average, window=window)
+    else:
+        kwargs.update(bandwidth=bandwidth, adaptive=adaptive,
+                      low_bias=low_bias, normalization=normalization)
+    raw.compute_psd(**kwargs)
+
+
+@requires_h5py
+@pytest.mark.parametrize('inst', ('raw', 'epochs', 'evoked'))
+def test_spectrum_io(inst, tmp_path, request):
+    """Test save/load of spectrum objects."""
+    inst = request.getfixturevalue(inst)
+    fname = tmp_path / 'spectrum.h5'
+    orig = inst.compute_psd()
+    orig.save(fname)
+    loaded = read_spectrum(fname)
+    assert orig == loaded
```

Comment:
@larsoner this test fails for evoked because of a limitation in pytest. The failure is:

```
_____________________________________________ test_spectrum_io[evoked] ______________________________________________
The requested fixture has no parameter defined for test:
    mne/time_frequency/tests/test_spectrum.py::test_spectrum_io[evoked]

Requested fixture '_evoked' defined in:
mne/conftest.py:306

Requested here:
/opt/miniforge3/envs/mnedev/lib/python3.10/site-packages/_pytest/fixtures.py:659
```

I can make it go away by removing `params=[testing._pytest_param()]` from the signature here:

https://github.com/mne-tools/mne-python/blob/2ba483ab91828be18a480e689d95f61c422ed6eb/mne/conftest.py#L305-L312

But presumably that will break your magic auto-skipping behavior.  The limitation is that `request.getfixturevalue()` cannot handle fixtures that have params, and this is [apparently by design](https://github.com/pytest-dev/pytest/issues/4666#issuecomment-456593913). A workaround is to use the [pytest-lazy-fixture](https://pypi.org/project/pytest-lazy-fixture) package.  Another alternative is to copy-paste a separate test for Evoked IO, like this:

```py
@requires_h5py
def test_evoked_spectrum_io(evoked, tmp_path):
    """Test save/load of spectrum objects."""
    fname = tmp_path / 'spectrum.h5'
    orig = evoked.compute_psd()
    orig.save(fname)
    loaded = read_spectrum(fname)
    assert orig == loaded
```
... which works because now the `evoked` fixture is listed directly in the test signature (instead of being accessed via `getfixturevalue()`), but is not very DRY.

What do you recommend?

Reply:
I would work around it with:
```
@requires_h5py
@pytest.mark.parametrize('inst', ('raw', 'epochs', 'evoked'))
def test_spectrum_io(inst, tmp_path, request, evoked):
    """Test save/load of spectrum objects."""
    # work around parametrized fixtures not being accessible
    # via request.getfixturevalue
    # https://github.com/pytest-dev/pytest/issues/4666#issuecomment-456593913  # noqa: E501
    if inst == 'evoked':
        inst = evoked
    else:
        inst = request.getfixturevalue(inst)
    fname = tmp_path / 'spectrum.h5'
    orig = inst.compute_psd()
    orig.save(fname)
    loaded = read_spectrum(fname)
    assert orig == loaded
```

Reply:
Ah, OK. Inelegant but practical. Will do.